### DISCUSSION TITLE
XText feature list and current state in langium

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#389] [General]
While playing with langium I encountered some small problems (e.g. when trying to define enum rules or using the boolean assignment operator `?=`). It would be great to have a feature list from XText and maybe some example demonstrating the ability of langium to implement this feature.

This could also be a community effort...

What do you think? Could such a list be integrated into the automated test suite and the examples? Or does this list already exist somewhere?

**Comment by msujew** [comment]
Hi @goto40,

that's a great idea! We could probably add a comparison page (Xtext vs. Langium) to our [langium-website](https://github.com/langium/langium-website). I don't think that this could be automated, but done as a community effort.

Regarding your mentioned issues:
1. We removed enum rules in favor of using data type rules which get generated as union types. E.g. a data type rule such as `State returns string: 'off' | 'on';` creates a generated ast type `export type State = 'off' | 'on'`.
2. Boolean assignment operators should work as they did in Xtext, do you mind sharing the issue you where experiencing with them?

**Reply by spoenemann** [reply]
Let's write a "Migration from Xtext" guide.

**Reply by trusktr** [reply]
Can you please write an Xtext vs Langium page for people learning to make a language from scratch (well, not totally from scratch but with help from a tool like Xtext or Langium), and wanting to evaluate their use case and which fits better?

I've been having fun with Langium! In what use cases do we want to use Xtext instead of Langium?

Perhaps it would be helpful to make these lists:

1. What's a list of things that Xtext can do that Langium can't?
2. Which of those things on the list is Langium planning to reach parity with, and which not?

**Reply by msujew** [reply]
Without trying to drag the Xtext framework into the dirt, but at this point there is very little reason to prefer Xtext over Langium for new projects. Except for one, very specific feature (model serialization), Langium already has full feature parity. The only use cases I can think of are:

* You want to stay in the Java ecosystem.
* You already use a platform like Eclipse RPC and want the best possible integration of your language into Eclipse.
* Your use case is very performance critical. Xtext doesn't do multi-processing too well by default, but it can be done. However, at this point it might make more sense to drop language engineering frameworks as a whole and roll your own solution instead. We have reached Xtext's performance limits often in the past.

As for the feature that's missing in Langium (model serialization, i.e. transforming an AST back into a textual representation), we plan to add that soon. We don't have a use case yet, that's why currently no one's working on that. But we still plan to work on that, just because we think it's a fairly useful feature for Langium to have.

**Comment by goto40** [comment]
Thank you very much for the prompt response!
(1) enums: I understand... your answer and example makes the thing clear!
(2) For boolean assignment operation it was my fault: I came from textx, where a boolean assignment implies the assignment to be optional... Reading through the specification made me understand I was wrong... I need an additional `?` to make the thing optional: `(opt?=opt)?`... Again, this was my fault. Very good work you did!

**Comment by goto40** [comment]
Is there a function generated by the langium code generator to check the validity of a string concerning a given union type from the grammar? E.g., `isState` in the example above.

**Reply by spoenemann** [reply]
I've seen this just now... I think it's a good idea, created #990 for it.
